source/connections-survive-step-down/tests/README.rst

```diff
@@ -0,0 +1,214 @@
+
+Adjust replica set configuration to have the following settings:
+
+- ``electionTimeoutMillis``: 1000
+
+Connect to each of the servers in the deployment directly and set the
+following server parameters:
+
+- ``enableElectionHandoff``: false
+
+Note: for the elections to work correctly, the values specified for
+replSetStepDown, secondaryCatchUpPeriodSecs and electionTimeoutMillis must
+satisfy the following relationship:
+
+``replSetStepDown`` >= ``secondaryCatchUpPeriodSecs`` + ``electionTimeoutMillis``/1000.0
+
+This ensures that each of the servers has time to vote in the election after
+the existing primary finishes waiting for other servers to catch up to it
+and before the existing primary's election black list time ends.
+
+For reference, the default values of the changed settings which should be
+reinstated once the tests are finished are as follows:
+
+- ``electionTimeoutMillis``: ``10000``
+- ``enableElectionHandoff``: ``true``
+- Replica set member ``priority``: 1
+
+Server Election Procedure
+-------------------------
+
+The following procedure should be used to elect a server N as the new primary,
+when another server E is the existing primary.
+
+1. Connect to the server E and retrieve current replica set configuration using
+   the `replSetGetConfig <https://docs.mongodb.com/manual/reference/command/replSetGetConfig/>`_
+   admin command.
+2. For each item in the ``members`` field of the configuration, examine the
+   value of the ``host`` field. If the value matches the address of the
+   server N, set ``priority`` of the respective item to 10. If the value
+   matches the address of the server E, set ``priority`` to 1. Otherwise
+   set ``priority`` to 0.
+3. Increment the ``version`` field of the replica set configuration.
+4. Reconfigure the replica set using the `replSetReconfig
+   <https://docs.mongodb.com/manual/reference/command/replSetReconfig/>`_
+   command, providing the modified configuration.
+5. Directly connect to the server N and execute `replSetFreeze
+   <https://docs.mongodb.com/manual/reference/command/replSetFreeze/>`_
+   admin command as follows: ``{replSetFreeze: 0}``. If this command
+   fails with operation failure code 95 ("cannot freeze node when primary or
+   running for election. state: Primary"), perform server selection to
+   discover the current primary. If the current primary is server N, stop
+   as the procedure is complete. Otherwise propagate the operation failure
+   error(*).
+6. Connect to the server E and execute the `replSetStepDown
+   <https://docs.mongodb.com/manual/reference/command/replSetStepDown/>`_
+   admin command as follows:
+   ``{replSetStepDown: 4, secondaryCatchUpPeriodSecs: 2}``.
+7. Connect directly to the server N and execute the following admin command:
+   ``{replSetStepUp: 1}``. If this command fails with an operation failure
+   error with code 125 ("Election failed"), repeat this step 7.
+8. Set the status of all servers in the topology to Unknown.
```

Comment:
For drivers that don't have direct access to SDAM internals (e.g. drivers built atop libmongoc), is there an alternative approach for this step? Perhaps by tweaking monitoring options for force a re-scan of the topology before server selection in the subsequent step?

Reply:
Yes, the exact wording of this step can be modified. The goal is to refresh the driver's topology view and find out the current primary.